### PR TITLE
Provide interface and individual namespaces for brew CLI commands

### PR DESCRIPTION
**Pull request by dduugg**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----
This is a strawman implementation of https://github.com/Homebrew/brew/issues/16814. It adds an `AbstractCommand`, which requires implementing `raw_args` (usually implemented as {$name}_args, `unparsed_args` could be a better name) and `run` (implemented as `${name}`). There's also a `CommandRegistry` to allow lookup by command name (used in brew.rb).

Two commands are ported as examples (one in `cmd`, one in `dev-cmd`). They are currently namespaced according to their current paths. (Best viewed with whitespace diffing disabled, due to indentation changes.)

(Tests forthcoming, pending consensus on approach/implementation.)

### CODE REVIEW

File: Library/Homebrew/cmd/list.rb
Comment (Bo98) [inline Library/Homebrew/cmd/list.rb:18]: I wonder if we can have some class-level DSL here so that we can do

```suggestion
      args do
```

Comment (MikeMcQuaid) [reply Library/Homebrew/cmd/list.rb:18]: or `cmd_args`

**Comment by MikeMcQuaid** [comment]
@dduugg I like the approach, nice work!

File: Library/Homebrew/cmd/list.rb
Comment (dduugg) [reply Library/Homebrew/cmd/list.rb]: Nice, I really like this suggestion, in particular for hiding the concept of a `CLI::Parser` in subclasses. PTAL at the latest commit.

File: Library/Homebrew/command_registry.rb
Comment (dduugg) [inline Library/Homebrew/command_registry.rb:5]: The existence of this module might be overkill, I could go either way. I originally had the registry implemented in `AbstractCommand`, but thought the registry polluted the concept of the superclass: https://github.com/Homebrew/brew/pull/16815/commits/4295a333d0051024677bf953413b492349573a20

**Comment by apainintheneck** [comment]
I'm not sure this is a great approach.

It seems like we have a base class with a callback to register all classes that inherit from it. Then, we use that registry to find and call the command class that matches the command the user entered. All of that makes sense.

The problem is that we currently don't load most of the commands. In fact, we only load one command per run 99% of the time. That allows us to lazy load a bunch of files based on the command that is getting run. If we have to load all commands into memory before running one of them, we'll be including most of the codebase by default which will likely negatively affect start up time.

I think we should preserve the current lazy loading behavior if possible.

**Comment by dduugg** [comment]
> I think we should preserve the current lazy loading behavior if possible.

There are no changes to the loading behavior, which is still a path search based on the command name. The registry is a convenience for mapping the command name to the class name, once the file has been loaded. I’m open to the critique that it’s over-designed for this use case. (I can take a look at returning the command class more directly as a result of the path resolution search.)

**Comment by apainintheneck** [comment]
Ah, okay. My bad. I completely misinterpreted what was being proposed here. I'm fine with this approach. It is unfortunate that we'll be up to three different command types but I guess that's the price we pay for backwards compatibility.

**Comment by dduugg** [comment]
@apainintheneck check out the latest commit. I've done away with the registry in favor of a simple search of `AbstactCommand`'s subclasses (using the new `subclasses` method [introduced](https://rubyreferences.github.io/rubychanges/3.1.html#classsubclasses) in Ruby 3.1). It nicely avoids the rubocop/sorbet issues associated with `const_get` or `ObjectSpace`. (It's also O(n), but i don't think that's an issue. 😸)

**Comment by apainintheneck** [comment]
> @apainintheneck check out the latest commit. I've done away with the registry in favor of a simple search of `AbstactCommand`'s subclasses (using the new `subclasses` method [introduced](https://rubyreferences.github.io/rubychanges/3.1.html#classsubclasses) in Ruby 3.1). It nicely avoids the rubocop/sorbet issues associated with `const_get` or `ObjectSpace`. (It's also O(n), but i don't think that's an issue. 😸)

This O(n) where n is most likely just 1, right? Or does n refer to all objects. I remember the Rails [#descendants](https://apidock.com/rails/Class/descendants) method being pretty inefficient because it had to walk the entire `ObjectSpace`. It's probably fine since we'll only call it once in the lifetime of the program.

I prefer this approach since we don't plan on referring to the registry more than once so we might as well just not build it in the first place. I'm fine with whichever you prefer though.

**Comment by dduugg** [comment]
> This O(n) where n is most likely just 1, right? Or does n refer to all objects. I remember the Rails [#descendants](https://apidock.com/rails/Class/descendants) method being pretty inefficient because it had to walk the entire `ObjectSpace`. It's probably fine since we'll only call it once in the lifetime of the program.
> 
> I prefer this approach since we don't plan on referring to the registry more than once so we might as well just not build it in the first place. I'm fine with whichever you prefer though.

Sorry for the lack of specificity, I was referring to the `command` lookup, which loops over `subclasses` now, instead of using a hash lookup. (I don't know the internals, though I suspect `subclasses` is much more efficient that what Rails had to work with). I expect them to be indistinguishable in practice.

**Comment by dduugg** [comment]
I believe initial feedback has been addressed, opening this for general review. (More than happy to port over more commands to demonstrate viability.)

File: Library/Homebrew/test/cmd/list_spec.rb
Comment (dduugg) [inline Library/Homebrew/test/cmd/list_spec.rb:8]: This doesn't play nicely with the new interface, but i don't think it's an especially useful test (it validates that `parse` can be called on the output of the command method, which has already called `parse` in nearly all cases). I think there are better tests in `abstract_command_spec`, including passing switches/flags.

**Comment by apainintheneck** [comment]
> I believe initial feedback has been addressed, opening this for general review. (More than happy to port over more commands to demonstrate viability.)

IMO this is probably enough unless there is a specific command that you think will be especially tricky to port over to the new interface. I think that shouldn't be the case though, right?

File: Library/Homebrew/test/cmd/list_spec.rb
Comment (apainintheneck) [reply Library/Homebrew/test/cmd/list_spec.rb:8]: I agree that it's not that useful. I would expect that using Sorbet here covers most of the same logic (all hail the type checker!). It might be useful to test that the contents of the `cmd_args` block is valid. That should be covered by integration tests though so I don't think we need to add anything extra. Maybe just something to keep in mind if a command doesn't have an integration test. I can't remember if they all have integration tests or not.

**Review by apainintheneck — APPROVED** [review]
This looks great to me.

File: Library/Homebrew/test/cmd/list_spec.rb
Comment (MikeMcQuaid) [reply Library/Homebrew/test/cmd/list_spec.rb:8]: @dduugg I disagree, this has caught legitimate issues for me in some tests that other tests have not (or other tests have been sufficiently opaque in their errors whereas this was very clear). Similarly, there's a few commands that can't really have good/any integration tests and this helps there too.

Would like to have this stick around in some form if possible! If not possible: let's talk/think about how we could have a test that does similar.

**Review by MikeMcQuaid — COMMENTED** [review]
Looks good to me except for one missing piece with `it_behaves_like "parseable arguments"`.

File: Library/Homebrew/test/cmd/list_spec.rb
Comment (dduugg) [inline Library/Homebrew/test/cmd/list_spec.rb:5]: Note that once we've ported all the commands, we can/should enable `RSpec/DescribeClass` (the commands are all currently in the `Homebrew` namespace, so the current approach makes sense).

File: Library/Homebrew/test/cmd/list_spec.rb
Comment (dduugg) [reply Library/Homebrew/test/cmd/list_spec.rb:8]: @MikeMcQuaid Thanks for the feedback. Let me know if the latest commit offers a viable replacement.

Methodology:

The `parseable arguments` test calls `#{command_name.tr("-", "_")}_args"` on `Homebrew`, in order to validate that the returned value responds to `parse` (I'm not sure why it does this, rather than check that it's an instance of `Homebrew::CLI::Parser.new`, but either way this would be better solved via `sig` enforcement, IMO)

In the new approach, `CLI::Parser` is called within the command initializer, where the parsing also happens. So, it would seem that validating that one can call `.new` on a command would be a stronger test than what was here before. I've also added the ability to pass a custom argv (since e.g. `prof` requires at least one argument).

**Review by apainintheneck — APPROVED** [review]

File: Library/Homebrew/cmd/list.rb
Comment (dduugg) [inline Library/Homebrew/cmd/list.rb:125]: I just realized that we no longer need to thread `args` through other methods (it's now an ivar, since each command is instance), so I've simplified some method signatures accordingly.

File: Library/Homebrew/cmd/list.rb
Comment (Bo98) [inline Library/Homebrew/cmd/list.rb:101]: Feels weaker from a typechecking point of view and feels weird seeing this but still seeing `args.formula?` and `args.verbose?`

Comment (apainintheneck) [reply Library/Homebrew/cmd/list.rb:101]: I'd overlooked this change. Is this needed to make Sorbet happy?

**Comment by dduugg** [comment]
> > I believe initial feedback has been addressed, opening this for general review. (More than happy to port over more commands to demonstrate viability.)
> 
> IMO this is probably enough unless there is a specific command that you think will be especially tricky to port over to the new interface. I think that shouldn't be the case though, right?

I've got a follow-up PR ready with ten commands ported and the only issue has been adding support for CamelCase to kebab-case mapping for multi-word commands.

File: Library/Homebrew/cmd/list.rb
Comment (dduugg) [reply Library/Homebrew/cmd/list.rb:101]: Correct. Because `CLI::Args` subclasses `OpenStruct`, many of the commands are just hash accessors, and not supported by Sorbet. Thus `args`, as the result of `Parser#parse`, has been treated as [untyped](https://github.com/Homebrew/brew/blob/ea2892f8/Library/Homebrew/cli/parser.rb#L311-L314):
```ruby
      # @return [Args] The actual return type is `Args`, but since `Args` uses `method_missing` to handle options, the
      #   `sig` annotates this as returning `T.untyped` to avoid spurious type errors.
      sig { params(argv: T::Array[String], ignore_invalid_options: T::Boolean).returns(T.untyped) }
      def parse(argv = ARGV.freeze, ignore_invalid_options: false)
```

As part of this new implementation, args is correctly typed, but may require changes:
```ruby
    # @note because `Args` makes use `OpenStruct`, subclasses may need to use a tapioca compiler,
    #   hash accessors, args.rbi, or other means to make this work with legacy commands:
    sig { returns(CLI::Args) }
    attr_reader :args
```
(Reviewers may have noticed that `AbstractCommand` has the rare `# typed: strong`, which is often overkill, but helpful for ancestors or other extensible tooling.)

Some `args` commands still work, for various reasons. As noted above, `args.rbi` is one option, but i don't especially love it (and I don't really have the bandwidth to write type signatures as part of the first pass of converting commands). It makes the methods available to all commands, rather than for just the command(s) that define the flag. I'm actively thinking about how to namespace the args as well, as a sort of v2 of what I'm introducing here, but I don't have anything concrete to propose at this time.

Comment (apainintheneck) [reply Library/Homebrew/cmd/list.rb:101]: The problem here is that while `args.pinned?` and `args[:pinned?]` are equivalent the new approach doesn't handle typos well. `args.pinnedd?` raises an error `Error: CLI arg for `pinnedd?` is not declared for this command` while `args[:pinnedd?]` doesn't fail and just returns nil.

Could we just keep cli args untyped and avoid this change? What additional type safety do we get by typing this attribute?

Comment (apainintheneck) [reply Library/Homebrew/cmd/list.rb:101]: Also, this is just another example of why a refactor of `CLI::Args` might be in the cards at some point.

Comment (dduugg) [reply Library/Homebrew/cmd/list.rb:101]: > The problem here is that while `args.pinned?` and `args[:pinned?]` are equivalent the new approach doesn't handle typos well. `args.pinnedd?` raises an error `Error: CLI arg for `pinnedd?` is not declared for this command` while `args[:pinnedd?]` doesn't fail and just returns nil.
> 
> Could we just keep cli args untyped and avoid this change? What additional type safety do we get by typing this attribute?

Correct, though it still represents a marginal type improvement anywhere this change is made (in that sorbet now recognizes the type of `args`). This means that `args.pinnedd?` at this call site previously would have passed typecheck. Again, there are alternatives if folks find this unacceptable (tapioca and args.rbi, in particular), or we could punt to when we want to refactor `CLI::Args` more broadly (though I'm partial to avoiding `T.untyped` in `AbstractCommand` from the outset).

Comment (Bo98) [reply Library/Homebrew/cmd/list.rb:101]: Sounds like we've moving from something that checked for errors at runtime so something that never checks for errors?

Comment (dduugg) [reply Library/Homebrew/cmd/list.rb:101]: > Sounds like we've moving from something that checked for errors at runtime so something that never checks for errors?

Could you elaborate, possibly with examples? I'm not sure that I understand this claim.

>  I'm not entirely sure what the marginal type improvement here is.

`args` was previously untyped by default, but now is typed. This is an issue accessing when OpenStruct hash values via methods.

Comment (Bo98) [reply Library/Homebrew/cmd/list.rb:101]:
```diff
diff --git a/Library/Homebrew/cmd/list.rb b/Library/Homebrew/cmd/list.rb
index 238f63f53a..d3f7953824 100644
--- a/Library/Homebrew/cmd/list.rb
+++ b/Library/Homebrew/cmd/list.rb
@@ -96,7 +96,7 @@ module Homebrew
 
           ls_args = []
           ls_args << "-1" if args[:"1?"]
-          ls_args << "-l" if args[:l?]
+          ls_args << "-l" if args.somethinginvalid?
           ls_args << "-r" if args[:r?]
           ls_args << "-t" if args[:t?]
```

results in:

```console
$ brew list
Error: CLI arg for `somethinginvalid?` is not declared for this command
```

---

However this:

```diff
diff --git a/Library/Homebrew/cmd/list.rb b/Library/Homebrew/cmd/list.rb
index 238f63f53a..f1fee19c14 100644
--- a/Library/Homebrew/cmd/list.rb
+++ b/Library/Homebrew/cmd/list.rb
@@ -96,7 +96,7 @@ module Homebrew
 
           ls_args = []
           ls_args << "-1" if args[:"1?"]
-          ls_args << "-l" if args[:l?]
+          ls_args << "-l" if args[:somethinginvalid?]
           ls_args << "-r" if args[:r?]
           ls_args << "-t" if args[:t?]
 
```

will result in no error.

Comment (dduugg) [reply Library/Homebrew/cmd/list.rb:101]: Perfect answer, thank you for the clear example. I believe the latest commit copies the behavior of method access to hash access, PTAL.

Comment (apainintheneck) [reply Library/Homebrew/cmd/list.rb:101]: I think that's a fair compromise for now. It'd be nice to make the syntax more consistent at some point but that will likely require a more involved refactor.

**Comment by dduugg** [comment]
> Looking great! One more thought for something to be improved here.
> 
> Given `CLI::Args` is a public API but there's talk about revamping it: perhaps something similar that doesn't use `OpenStruct` would be a better fit? I know it blows up the scope here a bit but just wanted to raise it may be a nicer solution.

Let's try a different approach. Let me see if I can whip up a tapioca compiler for `Homebrew::CLI::Args`, it'll make most of this discussion moot (`AbstractCommand` can remain strongly typed, and we can do whatever refactoring to CLI::Args that maintains the API as an independent effort).

Moving this to draft status until then. Thanks all for the thoughtful feedback, I super duper appreciate it. 🫶

**Comment by dduugg** [comment]
Opening this up for re-review, after merging https://github.com/Homebrew/brew/pull/16880

I've reverted the changes that used hash accessors for accessing args, because a) that style was considered less ergonomic and b) it has a different behavior than method access due to overrides in `Args` 🙀.

I've also reverted most of the changes to `Parser`, with the exception of making `option_to_name` a class method. (This is to avoid generating redundant type signatures for global args for every command).

File: Library/Homebrew/sorbet/rbi/dsl/homebrew/dev_cmd/prof.rbi
Comment (dduugg) [inline Library/Homebrew/sorbet/rbi/dsl/homebrew/dev_cmd/prof.rbi:9]: Cheating a bit here – command arg methods will be isolated to individual files per command, though (for now) arg methods all live in the same namespace (`Homebrew::CLI::Args`).

File: Library/Homebrew/sorbet/tapioca/compilers/args.rb
Comment (apainintheneck) [inline Library/Homebrew/sorbet/tapioca/compilers/args.rb:12]:
```suggestion
        Homebrew::CLI::Parser.global_options.each_with_object([]) do |(short_option, long_option), array|
          array << "#{Homebrew::CLI::Parser.option_to_name(short_option)}?"
          array << "#{Homebrew::CLI::Parser.option_to_name(long_option)}?"
        end.freeze,
```

Nit: This is a bit opaque. It's not obvious without looking at `.global_options` what the format could be for the block arguments. I don't have any suggestions that are equally succinct though. I have not tested the above code locally.

**Review by apainintheneck — APPROVED** [review]

File: Library/Homebrew/sorbet/tapioca/compilers/args.rb
Comment (dduugg) [reply Library/Homebrew/sorbet/tapioca/compilers/args.rb]: Great suggestion, PTAL at the next commit

Comment (apainintheneck) [reply Library/Homebrew/sorbet/tapioca/compilers/args.rb]: Much more readable, thanks!

**Review by MikeMcQuaid — COMMENTED** [review]
Looks good! Happy to have you self-merge when you're happy/ready @dduugg!

**Review by MikeMcQuaid — APPROVED** [review]

File: Library/Homebrew/sorbet/rbi/dsl/homebrew/dev_cmd/prof.rbi
Comment (MikeMcQuaid) [reply Library/Homebrew/sorbet/rbi/dsl/homebrew/dev_cmd/prof.rbi:9]: @dduugg Potentially outrageous suggestion that can wait to another PR (including discussion can wait) to not block this one:

What if all the per-command arguments were subclasses of `Homebrew::CLI::Args`?

Comment (dduugg) [reply Library/Homebrew/sorbet/rbi/dsl/homebrew/dev_cmd/prof.rbi:9]: That's the eventual idea, hence the "for now" above. 😉

**Pull request merged**
